Upgrade prometheus-cpp

## Enhancement

The current prometheus-cpp version is too old, with some bugs in compilation:

```
/home/yangkeao/Project/github.com/YangKeao/tiflash/contrib/prometheus-cpp/core/src/histogram.cc:53:38: error: no member named 'numeric_limits' in namespace 'std'
                              ? std::numeric_limits<double>::infinity()
                                ~~~~~^
/home/yangkeao/Project/github.com/YangKeao/tiflash/contrib/prometheus-cpp/core/src/histogram.cc:53:59: error: expected '(' for function-style cast or type construction
                              ? std::numeric_limits<double>::infinity()
                                                    ~~~~~~^
/home/yangkeao/Project/github.com/YangKeao/tiflash/contrib/prometheus-cpp/core/src/histogram.cc:53:62: error: no member named 'infinity' in the global namespace
                              ? std::numeric_limits<double>::infinity()
```

This has been fixed in https://github.com/jupp0r/prometheus-cpp/pull/442,  Jan 5, 2021. The current version of prometheus-cpp in tiflash is [ca1f34](https://github.com/jupp0r/prometheus-cpp/tree/ca1f3463e74d957d1cccddd4a1a29e3e5d34bd83) in Oct 25, 2019. 

This problem only occurs with libstdc++ >= 11 (as described in [fix](https://github.com/jupp0r/prometheus-cpp/pull/442).

--- a/contrib/prometheus-cpp-cmake/core/CMakeLists.txt
+++ b/contrib/prometheus-cpp-cmake/core/CMakeLists.txt
@@ -23,6 +23,12 @@ target_link_libraries(core
     $<$<AND:$<BOOL:UNIX>,$<NOT:$<BOOL:APPLE>>>:rt>
 )
 
+if(HAVE_CXX_LIBATOMIC)
+  # the exported library config must use libatomic unconditionally
+  # (the HAVE_CXX_LIBATOMIC variable should not leak into the target config)
+  target_link_libraries(core PUBLIC atomic)
+endif()
+
 target_include_directories(core
   PUBLIC
     $<BUILD_INTERFACE:${PROMETHEUS_SRC_DIR}/core/include>

--- a/contrib/prometheus-cpp-cmake/core/CMakeLists.txt
+++ b/contrib/prometheus-cpp-cmake/core/CMakeLists.txt
@@ -23,12 +23,6 @@ target_link_libraries(core
     $<$<AND:$<BOOL:UNIX>,$<NOT:$<BOOL:APPLE>>>:rt>
 )
 
-if(HAVE_CXX_LIBATOMIC)
-  # the exported library config must use libatomic unconditionally
-  # (the HAVE_CXX_LIBATOMIC variable should not leak into the target config)
-  target_link_libraries(core PUBLIC atomic)
-endif()
-
 target_include_directories(core
   PUBLIC
     $<BUILD_INTERFACE:${PROMETHEUS_SRC_DIR}/core/include>

--- a/contrib/prometheus-cpp-cmake/pull/CMakeLists.txt
+++ b/contrib/prometheus-cpp-cmake/pull/CMakeLists.txt
@@ -12,9 +12,18 @@ if(ENABLE_COMPRESSION)
 endif()
 
 add_library(pull
+  ${PROMETHEUS_SRC_DIR}/pull/src/basic_auth.cc
+  ${PROMETHEUS_SRC_DIR}/pull/src/basic_auth.h
+  ${PROMETHEUS_SRC_DIR}/pull/src/endpoint.cc
+  ${PROMETHEUS_SRC_DIR}/pull/src/endpoint.h
   ${PROMETHEUS_SRC_DIR}/pull/src/exposer.cc
   ${PROMETHEUS_SRC_DIR}/pull/src/handler.cc
   ${PROMETHEUS_SRC_DIR}/pull/src/handler.h
+  ${PROMETHEUS_SRC_DIR}/pull/src/metrics_collector.cc
+  ${PROMETHEUS_SRC_DIR}/pull/src/metrics_collector.h
+
+  ${PROMETHEUS_SRC_DIR}/pull/src/detail/base64.h
+
   $<$<BOOL:${USE_THIRDPARTY_LIBRARIES}>:$<TARGET_OBJECTS:civetweb>>
 )
 

--- a/contrib/prometheus-cpp-cmake/push/CMakeLists.txt
+++ b/contrib/prometheus-cpp-cmake/push/CMakeLists.txt
@@ -3,6 +3,8 @@ if(NOT CURL_FOUND)
 endif()
 
 add_library(push
+  ${PROMETHEUS_SRC_DIR}/push/src/curl_wrapper.cc
+  ${PROMETHEUS_SRC_DIR}/push/src/curl_wrapper.h
   ${PROMETHEUS_SRC_DIR}/push/src/gateway.cc
 )
 

